File: bf.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <stdbool.h>
#include <unistd.h>

#define MAX_PROG_SIZE 30000
#define MAX_CODE_SIZE 300000

void checkBounds(unsigned char *ptr, unsigned char *array)
{
    if (ptr < array || ptr >= array + MAX_PROG_SIZE)
    {
        fprintf(stderr, "Memory access out of bounds\n");
        exit(1);
    }
}

void compileBrainfuck(const char *bfCode, char *output)
{
    int dataPtr = 0;
    char *outPtr = output;

    outPtr += sprintf(outPtr, "#include <stdio.h>\n#include <stdlib.h>\n");

    outPtr += sprintf(outPtr, "void checkBounds(unsigned char *ptr, unsigned char *array) {if (ptr < array || ptr >= array + %d) {fprintf(stderr, \"Memory access out of bounds\\n\");exit(1);}}\n", MAX_PROG_SIZE);

    outPtr += sprintf(outPtr, "int main() {unsigned char array[%d] = {0};unsigned char *ptr = array;", MAX_PROG_SIZE);

    for (int i = 0; bfCode[i]; i++)
    {
        switch (bfCode[i])
        {
        case '>':
            outPtr += sprintf(outPtr, "++ptr;checkBounds(ptr, array);");
            break;
        case '<':
            outPtr += sprintf(outPtr, "--ptr;checkBounds(ptr, array);");
            break;
        case '+':
            outPtr += sprintf(outPtr, "++*ptr;");
            break;
        case '-':
            outPtr += sprintf(outPtr, "--*ptr;");
            break;
        case '.':
            outPtr += sprintf(outPtr, "putchar(*ptr);");
            break;
        case ',':
            outPtr += sprintf(outPtr, "*ptr=getchar();");
            break;
        case '[':
            outPtr += sprintf(outPtr, "while(*ptr) {");
            break;
        case ']':
            outPtr += sprintf(outPtr, "}");
            break;
        default:
            break;
        }
    }

    outPtr += sprintf(outPtr, "return 0;}");
}

void interpretBrainfuck(const char *bfCode)
{
    unsigned char array[MAX_PROG_SIZE] = {0};
    unsigned char *ptr = array;
    const char *codePtr = bfCode;
    int loopStack[MAX_PROG_SIZE];
    int loopStackPtr = 0;

    while (*codePtr)
    {
        checkBounds(ptr, array);

        switch (*codePtr)
        {
        case '>':
            ptr++;
            break;
        case '<':
            ptr--;
            break;
        case '+':
            ++(*ptr);
            break;
        case '-':
            --(*ptr);
            break;
        case '.':
            putchar(*ptr);
            break;
        case ',':
            *ptr = getchar();
            break;
        case '[':
            if (loopStackPtr >= MAX_PROG_SIZE)
            {
                fprintf(stderr, "Loop stack overflow\n");
                exit(1);
            }
            if (*ptr == 0)
            {
                int depth = 1;
                while (depth > 0)
                {
                    codePtr++;
                    if (*codePtr == '[')
                        depth++;
                    else if (*codePtr == ']')
                        depth--;
                }
            }
            else
            {
                loopStack[loopStackPtr++] = (int)(codePtr - bfCode);
            }
            break;
        case ']':
            if (loopStackPtr <= 0)
            {
                fprintf(stderr, "Loop stack underflow\n");
                exit(1);
            }
            if (*ptr != 0)
            {
                codePtr = bfCode + loopStack[loopStackPtr - 1];
            }
            else
            {
                loopStackPtr--;
            }
            break;
        }
        codePtr++;
    }
}

int main(int argc, char **argv)
{
    if (argc < 3)
    {
        fprintf(stderr, "Usage: %s [run|build] <brainfuck_file> [-s|--save]\n", argv[0]);
        return 1;
    }

    bool saveOutput = false;
    bool buildMode = false;

    if (argc > 3)
    {
        if (strcmp(argv[3], "-s") == 0 || strcmp(argv[3], "--save") == 0)
        {
            saveOutput = true;
        }
        else
        {
            fprintf(stderr, "Unknown option: %s\n", argv[3]);
            return 1;
        }
    }

    if (strcmp(argv[1], "build") == 0)
    {
        buildMode = true;
    }
    else if (strcmp(argv[1], "run") != 0)
    {
        fprintf(stderr, "Invalid mode: %s\n", argv[1]);
        return 1;
    }

    FILE *file = fopen(argv[2], "r");
    if (!file)
    {
        perror("Error opening file");
        return 1;
    }

    // Read the Brainfuck code
    char bfCode[MAX_PROG_SIZE];
    fread(bfCode, 1, MAX_PROG_SIZE - 1, file);
    bfCode[MAX_PROG_SIZE - 1] = '\0'; // Ensure null-termination
    fclose(file);

    if (buildMode)
    {
        char output[MAX_CODE_SIZE];
        compileBrainfuck(bfCode, output);

        FILE *outFile = fopen("output.c", "w");
        if (!outFile)
        {
            perror("Error creating output file");
            return 1;
        }
        fprintf(outFile, "%s", output);
        fclose(outFile);

        // Try compiling with Clang first, if not available, use GCC
        char compileCommand[256];
        snprintf(compileCommand, sizeof(compileCommand), "clang -o output output.c || gcc -o output output.c");
        int compileResult = system(compileCommand);

        if (compileResult != 0)
        {
            fprintf(stderr, "Compilation failed\n");
            if (!saveOutput)
            {
                unlink("output.c"); // Remove the temporary C file if compilation failed
            }
            return 1;
        }

        if (!saveOutput)
        {
            unlink("output.c"); // Remove the temporary C file if not saving
        }
    }
    else
    {
        interpretBrainfuck(bfCode);
    }

    return 0;
}

File: bf.rs
use std::fs::File;
use std::io::{self, Read, Write};
use std::path::Path;
use std::process::Command;

const MAX_PROG_SIZE: usize = 30_000;

fn check_bounds(ptr: usize, array: &[u8]) -> Result<(), String> {
    if ptr >= array.len() {
        Err("Memory access out of bounds".to_string())
    } else {
        Ok(())
    }
}

fn compile_brainfuck(bf_code: &str) -> String {
    let mut output = String::new();
    output.push_str("fn main() -> Result<(), Box<dyn std::error::Error>> {let mut array = [0u8; ");
    output.push_str(&MAX_PROG_SIZE.to_string());
    output.push_str("];let mut ptr: usize = 0;");
    let mut bracket_count = 0;
    for ch in bf_code.chars() {
        match ch {
            '>' => output.push_str("ptr += 1; check_bounds(ptr, &array)?;"),
            '<' => output.push_str("ptr = ptr.checked_sub(1).ok_or(\"Memory access out of bounds\")?; check_bounds(ptr, &array)?;"),
            '+' => output.push_str("array[ptr] = array[ptr].wrapping_add(1);"),
            '-' => output.push_str("array[ptr] = array[ptr].wrapping_sub(1);"),
            '.' => output.push_str("print!(\"{}\", array[ptr] as char);"),
            ',' => output.push_str("array[ptr] = io::stdin().bytes().next().and_then(|r| r.ok()).unwrap_or(0);"),
            '[' => {
                bracket_count += 1;
                output.push_str("while array[ptr] != 0 {");
            },
            ']' => {
                if bracket_count > 0 {
                    output.push_str("}");
                    bracket_count -= 1;
                }
            },
            _ => {}
        }
    }
    output.push_str("Ok(())}");
    output.push_str("fn check_bounds(ptr: usize, _array: &[u8]) -> Result<(), String> {if ptr >= ");
    output.push_str(&MAX_PROG_SIZE.to_string());
    output.push_str(" { Err(\"Memory access out of bounds\".to_string()) } else { Ok(()) }}");
    output
}

fn interpret_brainfuck(bf_code: &str) -> io::Result<()> {
    let mut array = [0u8; MAX_PROG_SIZE];
    let mut ptr: usize = 0;
    let mut code_ptr = 0;
    let mut loop_stack = Vec::new();

    while code_ptr < bf_code.len() {
        check_bounds(ptr, &array).map_err(|e| io::Error::new(io::ErrorKind::Other, e))?;

        match bf_code.chars().nth(code_ptr).unwrap() {
            '>' => ptr += 1,
            '<' => {
                ptr = ptr.checked_sub(1).ok_or(io::Error::new(
                    io::ErrorKind::Other,
                    "Memory access out of bounds",
                ))?
            }
            '+' => array[ptr] = array[ptr].wrapping_add(1),
            '-' => array[ptr] = array[ptr].wrapping_sub(1),
            '.' => print!("{}", array[ptr] as char),
            ',' => {
                let mut input = [0u8; 1];
                io::stdin().read_exact(&mut input)?;
                array[ptr] = input[0];
            }
            '[' => {
                if array[ptr] != 0 {
                    loop_stack.push(code_ptr);
                } else {
                    let mut depth = 1;
                    while depth > 0 {
                        code_ptr += 1;
                        match bf_code.chars().nth(code_ptr) {
                            Some('[') => depth += 1,
                            Some(']') => depth -= 1,
                            None => {
                                return Err(io::Error::new(
                                    io::ErrorKind::Other,
                                    "Mismatched brackets",
                                ))
                            }
                            _ => {}
                        }
                    }
                }
            }
            ']' => {
                if array[ptr] != 0 {
                    code_ptr = *loop_stack
                        .last()
                        .ok_or(io::Error::new(io::ErrorKind::Other, "Mismatched brackets"))?;
                } else {
                    loop_stack.pop();
                }
            }
            _ => {}
        }
        code_ptr += 1;
    }
    Ok(())
}

fn main() -> io::Result<()> {
    let args: Vec<String> = std::env::args().collect();
    if args.len() < 3 {
        eprintln!(
            "Usage: {} [run|build] <brainfuck_file> [-s|--save]",
            args[0]
        );
        return Err(io::Error::new(
            io::ErrorKind::InvalidInput,
            "Incorrect number of arguments",
        ));
    }

    let build_mode = args[1] == "build";
    let save_output = args.len() > 3 && (args[3] == "-s" || args[3] == "--save");

    let bf_path = Path::new(&args[2]);
    let mut bf_file = File::open(bf_path)?;
    let mut bf_code = String::new();
    bf_file.read_to_string(&mut bf_code)?;

    if build_mode {
        let rust_code = compile_brainfuck(&bf_code);
        let mut out_file = File::create("output.rs")?;
        out_file.write_all(rust_code.as_bytes())?;

        let status = Command::new("rustc")
            .arg("output.rs")
            .arg("-o")
            .arg("output")
            .status()?;

        if status.success() {
            if !save_output {
                std::fs::remove_file("output.rs")?;
            }
        } else {
            return Err(io::Error::new(io::ErrorKind::Other, "Compilation failed"));
        }
    } else {
        interpret_brainfuck(&bf_code)?;
    }

    Ok(())
}
